# Example notebook

We publish our results as `Pandas` dataframes that can be readily queried. Below we show how to access these, but point readers to the extensive documentation for more information on how to use them: https://pandas.pydata.org/docs/index.html.

In [1]:
import os
import pandas as pd

# Change working directory of the notebook to access the Python code
os.chdir('../python')

Initialising the data to pass to the dataframes can take a little while. On my machine the next cell takes about 5 mins to run.

In [2]:
%%time

from database import get_tree_level_dataframe, get_loop_level_dataframe

CPU times: user 5min 12s, sys: 623 ms, total: 5min 13s
Wall time: 5min 17s


Call these functions to return the dataframes. For example

In [3]:
df = get_loop_level_dataframe()

The dataframe contains many keys. These are summarised below.

| Key                         | Description                                                                               |
|-----------------------------|-------------------------------------------------------------------------------------------|
| `smeft_op`                  | The `sympy` symbol representing the coefficient of the SMEFT operator being matched on to |
| `smeft_label`               | The label of the SMEFT operator                                                           |
| `smeft_flavour`             | The flavour indices of the SMEFT operator                                                 |
| `smeft_op_expr`             | The symbolic expression implementing the SMEFT-LEFT matching provided in the paper        |
| `process`                   | The decay process generated in LaTeX                                                      |
| `gamma`                     | The symbolic decay rate                                                                   |
| `left_dimension`            | The dimension of the LEFT operator generated                                              |
| `gamma_coeff_1`             | The decay rate with the dimensionless coefficient set to 1                                |
| `left_op`                   | The label of the LEFT operator driving the decay process                                  |
| `left_flavour`              | The flavour indices of the LEFT operator                                                  |
| `lambda_limit`              | The symbolic lower limit on the underlying new-physics scale                              |
| `lambda_limit_coeff_1`      | The numeric lower limit on the underlying new-physics scale with coefficient set to 1     |
| `lifetime_limit`            | The experimental lifetime limit in years                                                  |
| `lifetime_limit_ref`        | The reference from which the experimental bound is taken                                  |
| `dim`                       | A copy of `left_dimension`                                                                |
| `matrix_elem`               | The nuclear matrix element relevant to the nucleon decay calculation                      |
| `fieldstring_label`         | The label of the field-string operator                                                    |
| `fieldstring_flavour`       | The flavour indices of the field string                                                   |
| `latex`                     | The LaTeX expression of the matching estimate                                             |
| `gamma_fieldstring_coeff`   | The decay rate with the matching estimate pasted in for the SMEFT coefficient             |
| `gamma_fieldstring_coeff_1` | Same as `gamma_fieldstring_coeff` but for field-string coefficient set to 1               |


In [4]:
df.keys()

Index(['smeft_op', 'smeft_label', 'smeft_flavour', 'smeft_op_expr', 'process',
       'gamma', 'left_dimension', 'gamma_coeff_1', 'left_op', 'left_flavour',
       'lambda_limit', 'lambda_limit_coeff_1', 'lifetime_limit',
       'lifetime_limit_ref', 'dim', 'matrix_elem', 'fieldstring_label',
       'fieldstring_flavour', 'latex', 'gamma_fieldstring_coeff',
       'gamma_fieldstring_coeff_1'],
      dtype='object')

This dataframe can be queried in the usual way. For example, suppose we want to see contributions to $p \to \pi^{0} e^{+}$ UV models generating operator 11 at tree level

In [8]:
df[(df.fieldstring_label == "11") & (df.process == "p \\to \\pi^{0} e^{+}")]

smeft_op smeft_label smeft_flavour  \
0    K_duql_0_0_0_0        duql          1111   
8    K_duql_0_0_0_0        duql          1111   
37   C_qqql_0_0_0_0        qqql          1111   
38   C_qqql_0_0_0_0        qqql          1111   
39   C_qqql_0_0_0_0        qqql          1111   
177  C_qqql_0_1_0_0        qqql          1211   
178  C_qqql_0_1_0_0        qqql          1211   
179  C_qqql_0_1_0_0        qqql          1211   
237  C_qqql_0_2_0_0        qqql          1311   
238  C_qqql_0_2_0_0        qqql          1311   
239  C_qqql_0_2_0_0        qqql          1311   

                                         smeft_op_expr              process  \
0                                       K_duql_0_0_0_0  p \to \pi^{0} e^{+}   
8                                       K_duql_0_0_0_0  p \to \pi^{0} e^{+}   
37                            0.9746875*C_qqql_0_0_0_0  p \to \pi^{0} e^{+}   
38                            0.9746875*C_qqql_0_0_0_0  p \to \pi^{0} e^{+}   
39                            0.9746875*C_qqql_0_0_0_0  p \to \pi^{0} e^{+}   
177                               0.225*C_qqql_0_1_0_0  p \to \pi^{0} e^{+}   
178                               0.225*C_qqql_0_1_0_0  p \to \pi^{0} e^{+}   
179                               0.225*C_qqql_0_1_0_0  p \to \pi^{0} e^{+}   
237  C_qqql_0_2_0_0*(0.00149597634375 - 0.003274212...  p \to \pi^{0} e^{+}   
238  C_qqql_0_2_0_0*(0.00149597634375 - 0.003274212...  p \to \pi^{0} e^{+}   
239  C_qqql_0_2_0_0*(0.00149597634375 - 0.003274212...  p \to \pi^{0} e^{+}   

                                                 gamma  left_dimension  \
0    0.000452573724054281*Abs(K_duql_0_0_0_0/LAMBDA...               6   
8    0.000452573724054281*Abs(K_duql_0_0_0_0/LAMBDA...               6   
37   0.000387774971411024*Abs(C_qqql_0_0_0_0/LAMBDA...               6   
38   0.000387774971411024*Abs(C_qqql_0_0_0_0/LAMBDA...               6   
39   0.000387774971411024*Abs(C_qqql_0_0_0_0/LAMBDA...               6   
177  2.06639821421001e-5*Abs(C_qqql_0_1_0_0/LAMBDA*...               6   
178  2.06639821421001e-5*Abs(C_qqql_0_1_0_0/LAMBDA*...               6   
179  2.06639821421001e-5*Abs(C_qqql_0_1_0_0/LAMBDA*...               6   
237  5.28933120085494e-9*Abs(C_qqql_0_2_0_0/LAMBDA*...               6   
238  5.28933120085494e-9*Abs(C_qqql_0_2_0_0/LAMBDA*...               6   
239  5.28933120085494e-9*Abs(C_qqql_0_2_0_0/LAMBDA*...               6   

                      gamma_coeff_1   left_op left_flavour  ...  \
0    0.000452573724054281/LAMBDA**4  S,RL_duu         1111  ...   
8    0.000452573724054281/LAMBDA**4  S,RL_duu         1111  ...   
37   0.000387774971411024/LAMBDA**4  S,LL_duu         1111  ...   
38   0.000387774971411024/LAMBDA**4  S,LL_duu         1111  ...   
39   0.000387774971411024/LAMBDA**4  S,LL_duu         1111  ...   
177   2.06639821421001e-5/LAMBDA**4  S,LL_duu         1111  ...   
178   2.06639821421001e-5/LAMBDA**4  S,LL_duu         1111  ...   
179   2.06639821421001e-5/LAMBDA**4  S,LL_duu         1111  ...   
237   5.28933120085494e-9/LAMBDA**4  S,LL_duu         1111  ...   
238   5.28933120085494e-9/LAMBDA**4  S,LL_duu         1111  ...   
239   5.28933120085494e-9/LAMBDA**4  S,LL_duu         1111  ...   

           lifetime_limit_ref dim  matrix_elem fieldstring_label  \
0    Super-Kamiokande:2020wjk   6     0.224860                11   
8    Super-Kamiokande:2020wjk   6     0.224860                11   
37   Super-Kamiokande:2020wjk   6    -0.213546                11   
38   Super-Kamiokande:2020wjk   6    -0.213546                11   
39   Super-Kamiokande:2020wjk   6    -0.213546                11   
177  Super-Kamiokande:2020wjk   6    -0.213546                11   
178  Super-Kamiokande:2020wjk   6    -0.213546                11   
179  Super-Kamiokande:2020wjk   6    -0.213546                11   
237  Super-Kamiokande:2020wjk   6    -0.213546                11   
238  Super-Kamiokande:2020wjk   6    -0.213546                11   
239  Super-Kamiokande:2020wjk  

What lower bound can we set on the scale $\Lambda$ characterising these models, for unit operator coefficient?

In [13]:
df[(df.fieldstring_label == "11") & (df.process == "p \\to \\pi^{0} e^{+}")].sort_values(by="lambda_limit_coeff_1", ascending=False).iloc[0]

smeft_op                                                        C_qqql_0_0_0_0
smeft_label                                                               qqql
smeft_flavour                                                             1111
smeft_op_expr                                         0.9746875*C_qqql_0_0_0_0
process                                                    p \to \pi^{0} e^{+}
gamma                        0.000387774971411024*Abs(C_qqql_0_0_0_0/LAMBDA...
left_dimension                                                               6
gamma_coeff_1                                   0.000387774971411024/LAMBDA**4
left_op                                                               S,LL_duu
left_flavour                                                              1111
lambda_limit                         1873243117644.49*sqrt(Abs(C_11,_0_0_0_1))
lambda_limit_coeff_1                                          1873243117644.49
lifetime_limit                           23999999999

And we see that the best bound from $p \to \pi^{0} e^{+}$ on models generating operator 16 at tree level is 1873243117644.49. The bound is on the field-string flavour indices 1112. The bound saturation is 0.42, which shows that a better limit can be set from another decay modes.

In the paper, we present the bound saturation for different flavours of the field-string operator 16. Here we reproduce that analysis.

First calculate the bound saturation

In [5]:
# Import LAMBDA to set later
from tables import LAMBDA

def calc_exp_ratio(row):
    """Function to calculate the ratio of the decay rate to its experimental bound."""
    inv_gev_per_year = 7.625e30
    value_in_inv_gev = lambda x: inv_gev_per_year * x

    # Set LAMBDA to 1 since it cancels in the full ratio anyway
    gamma = row.gamma_fieldstring_coeff_1.subs({LAMBDA: 1.})
    gamma_limit = 1.0 / value_in_inv_gev(row.lifetime_limit)

    ratio = gamma / gamma_limit
    return ratio

df["exp_ratio"] = df.apply(calc_exp_ratio, axis=1)
# A dictionary mapping a tuple (fieldstring_label, fieldstring_flavour) to a float (the max ratio of the decay rate to the experimental limit)
max_ratio_dict = df[["fieldstring_label", "fieldstring_flavour", "exp_ratio"]].groupby(by=["fieldstring_label", "fieldstring_flavour"]).max().to_dict()["exp_ratio"]

def calc_bound_saturation(row):
    """Calculate the double ratio: exp_ratio / max(exp_ratio)."""
    return row.exp_ratio / max_ratio_dict[(row.fieldstring_label, row.fieldstring_flavour)]

df["bound_saturation"] = df.apply(calc_bound_saturation, axis=1)

We can query the dataframe for just those rows that are related to operator 16, and create restructure the table to look like the correlation plots presented in the paper.

In [20]:
pd.pivot_table(
    df[df.fieldstring_label == '16'],
    values='bound_saturation',
    index=['fieldstring_label', 'fieldstring_flavour'],
    columns=['process'],
    aggfunc=lambda x: float(max(x)), 
    fill_value=0,
)

process                                n \to K^{0} \nu  n \to \eta^{0} \nu  \
fieldstring_label fieldstring_flavour                                        
16                1111                        0.000132            0.000756   
                  1112                        0.019496            0.001273   
                  1113                        0.019496            0.001209   
                  1121                        0.019496            0.001273   
                  1122                        0.019496            0.001273   
                  1123                        0.019496            0.001209   
                  1131                        0.019496            0.001209   
                  1132                        0.019496            0.001209   
                  1133                        0.039613            0.000673   
                  1211                        0.019496            0.001273   
                  1212                        0.019496            0.001273   
                  1213                        0.022282            0.000302   
                  1221                        0.019496            0.001273   
                  1222                        0.019496            0.001273   
                  1223                        0.022282            0.000302   
                  1231                        0.022282            0.000302   
                  1232                        0.022282            0.000302   
                  1233                        0.022282            0.000302   
                  1311                        0.019496            0.000177   
                  1312                        0.022282            0.000177   
                  1313                        0.022282            0.000177   
                  1321                        0.022282            0.000177   
                  1322                        0.022282            0.000177   
                  1323                        0.022282            0.000177   
                  1331                        0.022282            0.000177   
                  1332                        0.022282            0.000177   
                  1333                        0.022282            0.000177   

process                                n \to \pi^{-} e^{+}  n \to \pi^{0} \nu  \
fieldstring_label fieldstring_flavour                                           
16                1111                            0.110259           0.043607   
                  1112                            0.046390           0.073389   
                  1113                            0.044071           0.069720   
                  1121                            0.046390           0.073389   
                  1122                            0.046390           0.073389   
                  1123                            0.044071           0.069720   
                  1131                            0.044071           0.069720   
                  1132                            0.044071           0.069720   
                  1133                            0.024516           0.038784   
                  1211                            0.000000           0.073389   
                  1212                            0.000000           0.073389   
                  1213                            0.000000           0.017430   
                  1221                            0.000000           0.073389   
                  1222                            0.000000           0.073389   
                  1223                            0.000000           0.017430   
                  1231                            0.000000           0.017430   
                  1232                            0.000000           0.017430   
                  1233                            0.000000           0.017430   
                  1311                            0.000000           0.010206   
                  1312              